### PR TITLE
docs: Distill r0bre's 100 Solana tips into the dev skill

### PR DESCRIPTION
**Pull request by gitteri**
Pulls the durable lessons from r0bre's 100 Daily Solana Tips (accretionxyz) into the skill, routed to where each fits and deduped against what the skill already covers.

What's here:

New `program-design-patterns.md` reference: state layout, PDA seed conventions, account parallelization, account lifecycle, a consolidated CPI limits table, the flashloan/in-transaction-credit pattern via Instructions-sysvar introspection, naming conventions, explicit state machines, and Rust/Anchor ergonomics (lifetimes, macros, newtypes). Registered in SKILL.md progressive disclosure.

`security.md`: adds field-tested attack vectors not already in the core categories (unvalidated remaining_accounts, self-reentrancy, log injection, slot/epoch boundaries, TOCTOU, pool squatting, donation attacks, on-chain randomness, rounding direction, unchecked casts, upgradeable-dep risk, unsafe blocks, frontrunning, malicious RPC, stale state around CPIs, unsafe arbitrary invoke, transient owner spoofing, backdoors), plus matching review questions.

`testing.md`: a fuzzing section (Trident, LibFuzzer).

`resources.md`: Trident, cargo-expand, sbpf, and an ECC primer link.

Deliberately skipped tips already fully covered (e.g. the core 9 vuln categories, existing test frameworks) to avoid duplication.

Source: https://accretionxyz.substack.com/p/r0bres-100-daily-solana-tips

### CODE REVIEW

File: skills/solana-dev/SKILL.md [hunk 1]
```diff
@@ -173,6 +173,7 @@ Surfpool also ships its own MCP server (`surfpool mcp`, stdio) for driving local
 - Legacy web3.js routing (v3 status + migration skill): [kit-web3-interop.md](references/kit-web3-interop.md)
 - Anchor programs: [programs/anchor.md](references/programs/anchor.md)
 - Pinocchio programs: [programs/pinocchio.md](references/programs/pinocchio.md)
+- Program design patterns (state layout, PDAs, parallelization, cranks, ergonomics): [program-design-patterns.md](references/program-design-patterns.md)
```
[the remark this answers is not in the document]
Comment (gitteri) [reply skills/solana-dev/SKILL.md]: Addressed: the branch is now a single commit (3b36b5b) that shows as verified/signed, so the provenance requirement is satisfied. Resolving.

File: skills/solana-dev/references/testing.md [hunk 2]
```diff
@@ -388,6 +388,15 @@ NO_DNA=1 surfpool start --legacy-anchor-compatibility --anchor-test-config-path
 
 Keep a small suite that runs against devnet before releases: deploy, exercise one happy path per instruction, verify explorer-visible effects. Use Kit with `solanaRpc({ rpcUrl })` pointed at devnet and a funded keypair via `signerFromFile('~/.config/solana/id.json')`. These are slow and flaky by nature — never gate PRs on them.
 
+## Fuzz Testing
+
+Fuzzing generates large volumes of randomized inputs and programmatically asserts the program still behaves correctly — surfacing edge cases, logic errors, and economic attack vectors that hand-written tests miss. Solana programs take two input surfaces: **instruction data** (easy to vary within constraints) and **accounts** (the hard part — you must synthesize valid account structures with varied ownership, balances, and data layouts).
+
+- **[Trident](https://ackee.xyz/trident/docs/latest/)** (Ackee) — the dedicated Solana fuzzing framework: generates instruction sequences targeting potential vulnerabilities, with account-state modeling built in. Start here for program-level fuzzing.
```
Comment (dev-jodee) [inline skills/solana-dev/references/testing.md:395]: I've actually had pleasant experiences with https://github.com/asymmetric-research/crucible as well, especially with litesvm, might be good to add

File: skills/solana-dev/references/programs/design-patterns.md [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+  ```
+- **Understand the Context lifetimes (`'a, 'b, 'c, 'info`).** They're relative lifetimes on `Context`'s reference fields (`program_id`, `accounts`, `remaining_accounts`). A bound like `'c: 'info` means "`'c` lives at least as long as `'info`" — the `remaining_accounts` reference can't outlive the `AccountInfo` data it points into. `'info` is the same lifetime used across your `#[derive(Accounts)]` struct.
+- **Write your own macros to kill copy-paste bugs.** Repeated fee math / token transfers / safe-math copied 10× is where copy-paste bugs live. A single macro (declarative, derive, or attribute — e.g. an `admin_only` constraint, a CU-logging wrapper, an account-size derive) gives you one sound implementation to reuse. Auditors: `cargo expand` rolls macros out to real code for review.
+
+## Vault topology
+
+- **Unified vault** (one global PDA for all deposits): simple TVL, but concentrates risk and creates a write-lock chokepoint.
+- **Multi-vault** (per-pool or per-user): isolates risk and parallelizes writes, at the cost of more complex TVL aggregation. The right choice depends on your parallelism and risk-isolation needs.
```
Comment (dev-jodee) [inline skills/solana-dev/references/programs/design-patterns.md]: Should we mention at the cost of complexity but also increase rent cost for users? (at least until we get rent reduct)

File: skills/solana-dev/references/concepts.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+```bash
+solana rent 165   # lamports needed to make a 165-byte account rent-exempt
+```
+
+## Keys and off-curve PDAs
+
+Solana uses Ed25519. A private key is a secret scalar `k`; the public key is `k·P`, a point on the curve — cheap to compute forward, infeasible to invert.
+
+**PDAs are addresses deliberately placed *off* the curve.** No scalar maps to them, so they have no private key and can never be signed for by a keypair. Only the owning program can authorize them, by supplying the seeds. That is the whole basis of program-controlled accounts.
```
Comment (dev-jodee) [inline skills/solana-dev/references/concepts.md:22]: Should we mention to ALWAYS validate pda addresses by validating that the owning program is your program.

**Review by dev-jodee — COMMENTED** [review]
looks good, couple of suggestions

**Review by amilz — APPROVED** [review]

**Review by dev-jodee — APPROVED** [review]
Documentation distillation of r0bre's 100 Solana Tips. Covers program design patterns, security vectors, testing/fuzzing, and runtime concepts. Greptile checks pass. Good to merge.
